src/utils.jl

```diff
@@ -66,3 +66,88 @@ function check_pkg_version(pkg::String, ver::String;
            pkg_info[pkg].version >= VersionNumber(ver) :
            pkg_info[pkg].version > VersionNumber(ver)
 end
+
+
+# RetCode handling for BBO and others.
+using Logging
+ # Define the ReturnCode type using @enum
+@enum ReturnCode begin
+    Default
+    Success
+    Terminated
+    MaxIters
+    MaxTime
+    DtLessThanMin
+    Unstable
+    InitialFailure
+    ConvergenceFailure
+    Failure
+    Infeasible
+end
+ 
+# Define a dictionary to map regular expressions to ReturnCode values
+const STOP_REASON_MAP = Dict(
+    r"Delta fitness .* below tolerance .*" => ReturnCode.Success,
+    r"Fitness .* within tolerance .* of optimum" => ReturnCode.Success,
+    r"Terminated" => ReturnCode.Terminated,
+    r"MaxIters|MAXITERS_EXCEED" => ReturnCode.MaxIters,
+    r"MaxTime|TIME_LIMIT" => ReturnCode.MaxTime,
+    r"DtLessThanMin" => ReturnCode.DtLessThanMin,
+    r"Unstable" => ReturnCode.Unstable,
+    r"InitialFailure" => ReturnCode.InitialFailure,
+    r"ConvergenceFailure|ITERATION_LIMIT" => ReturnCode.ConvergenceFailure,
+    r"Infeasible|INFEASIBLE|DUAL_INFEASIBLE|LOCALLY_INFEASIBLE|INFEASIBLE_OR_UNBOUNDED" => ReturnCode.Infeasible
+)
+
+# Function to deduce ReturnCode from a stop_reason string using the dictionary
+function deduce_retcode(stop_reason::String)::ReturnCode
+    for (pattern, retcode) in STOP_REASON_MAP
+        if occursin(pattern, stop_reason)
+        	return retcode
+        end
+    end
+    @warn "Unrecognized stop reason: $stop_reason. Defaulting to ReturnCode.Failure."
+    return ReturnCode.Failure
+end
+
+function Base.convert(::Type{ReturnCode.T}, retcode::Union{Symbol, String})
```

Comment:
This might be useful, but where do you use it? Is there somewhere in the package we are still returning Symbols iirc we had moved to only SciML retcodes a while back and there weren't any left to update

Reply:
This the convert function which was updated in BBO and it can handle String as well as symbol so if we need to use the convert function where a symbol is used there also it will work.

Reply:
I think I had missed that it was dispatching on the union. This is nice.

Let's do it like this though, to make it use dispatch so the you don't need the branches

```
function Base.convert(::Type{ReturnCode.T}, retcode::Symbol)
#all the symbol checks
end

function Base.convert(::Type{ReturnCode.T}, retcode::String)
#same as deduce_retcode
end
```

but one issue is there's already a convert method here https://github.com/SciML/SciMLBase.jl/blob/master/src/retcodes.jl#L407 so we can make something else that's just used in SciMLBase, convert methods are more useful when you want to allow automatic type conversions between two types what we want to do here is more of deduction so I like `deduce_retcode` let's use that

Reply:
We are still using the convert function in SciMLBase but this convert reduces any Failure returns due to unrecognised symbols by only passing selective retcode options like an extra layer of checking, and the union reduces code length.

Reply:
This will overwrite the method from scimlbase that's not kosher to do